### PR TITLE
Python runtime metrics

### PR DESCRIPTION
**Pull request by sjaiswalvmware**
### Description:
Changes for Python run time metrics.

### Review info:
Jira ticket: https://wavefront.atlassian.net/browse/INT-434
Reviewers: @ustinov @haosong @vikramraman

### CODE REVIEW

File: wavefront_pyformance/runtime_metrics.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import resource
+import threading
+
+import psutil
+
+
+class RuntimeCollector(object):
+    """Python Runtime Metrics Collection Class."""
+
+    def __init__(self, registry=None):
```
Comment (sushantdewan123) [inline wavefront_pyformance/runtime_metrics.py:16]: inside init method
you should report hearbeat metrics
see example - 
https://github.com/wavefrontHQ/wavefront-runtime-sdk-jvm/blob/32f51d3b548a937f72905a85efe245090bcd7c61/src/main/java/com/wavefront/sdk/appagent/jvm/reporter/WavefrontJvmReporter.java#L42

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: wavefront_pyformance/wavefront_reporter.py [hunk 2]
```diff
@@ -23,7 +24,8 @@ class WavefrontReporter(pyformance.reporters.reporter.Reporter):
 
     # pylint: disable=too-many-arguments
     def __init__(self, source='wavefront-pyformance', registry=None,
-                 reporting_interval=10, clock=None, prefix='', tags=None):
+                 reporting_interval=10, clock=None, prefix='', tags=None,
```
Comment (sushantdewan123) [inline wavefront_pyformance/wavefront_reporter.py:27]: default to 60 seconds?

**Review by sushantdewan123 — CHANGES_REQUESTED** [review]

File: example_runtime_metrics.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+def report_metrics(host, server='', token=''):
+    """Runtime Metric Reporting Function Example."""
+    reg = tagged_registry.TaggedRegistry()
+    wf_proxy_reporter = wavefront_reporter.WavefrontProxyReporter(
+        host=host, port=2878, registry=reg,
+        source='runtime-metric-test',
+        tags={'global_tag1': 'val1', 'global_tag2': 'val2'},
+        prefix='python.runtime.',
+        runtime_metric=True).report_minute_distribution()
+    # wf_direct_reporter = wavefront_reporter.WavefrontDirectReporter(
```
Comment (ustinov) [inline example_runtime_metrics.py]: Any reason why this is commented out?

File: wavefront_pyformance/runtime_metrics.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
+"""Python Runtime Metrics Collection Class."""
+# pylint: disable=E0012,R0205
```
Comment (ustinov) [inline wavefront_pyformance/runtime_metrics.py]: This should be next to the class declaration

File: README.md [hunk 5]
```diff
@@ -84,3 +84,16 @@ reg = pyformance.MetricsRegistry()
 h_0 = wavefront_histogram.wavefront_histogram(reg, 'requests_duration')
 h_0.add(10)
 ```
+
+### Python Runtime Metrics
+
+To enable Python runtime metrics reporting, set the `runtime_metric` flag to True:
```
Comment (ustinov) [inline README.md]: Shouldn't the flag name be `runtime_metrics`, WDYT?

File: .travis.yml [hunk 6]
```diff
@@ -12,6 +12,7 @@ install:
 - pip install flake8 flake8-colors flake8-docstrings flake8-import-order
 - pip install pep8-naming
 - pip install pylint
+- pip install psutil
```
Comment (ustinov) [inline .travis.yml]: Please also add `psutil` as a requirement in `setup.py` file

**Review by ustinov — CHANGES_REQUESTED** [review]

**Comment by ustinov** [comment]
@sjaiswalvmware here's a brief test I did:

1) Initialized and activated a new virtual environment
```
$ python3.8 -m venv ~/venv/38-wfpyformance-test
$ . ~/venv/38-wfpyformance-test/bin/activate
$ pip install -U --upgrade-strategy=eager pip setuptools
```

2) Installed the updated `wavefront-pyformance` from your branch
```
$ pip install -U --upgrade-strategy=eager git+https://github.com/sjaiswalvmware/wavefront-pyformance@sjaiswal/python-runtime-metrics
```

3) Tried to run the `example.py` and it failed because `psutil` dependency was missing (see my comment and request for a change above)

4) Installed the `psutil` manually by `pip install psutil`

5) Tried running the `./example_runtime_metrics.py` but it turned out to miss `+x` permission

6) Tried running the `example_runtime_metrics.py` as `python example_runtime_metrics.py`
```
$ python example_runtime_metrics.py 

usage: example_runtime_metrics.py [-h] host
example_runtime_metrics.py: error: the following arguments are required: host
```
which is ok

7) Tried running the `python example_runtime_metrics.py` providing a host name
```
$ python example_runtime_metrics.py qa4
Traceback (most recent call last):
  File "example_runtime_metrics.py", line 41, in <module>
    report_metrics(ARGS.host)
  File "example_runtime_metrics.py", line 27, in report_metrics
    wf_proxy_reporter.report_now()
  File "/Users/austinov/wavefronthq/wavefront-pyformance/wavefront_pyformance/wavefront_reporter.py", line 134, in report_now
    super(WavefrontProxyReporter, self).report_now(registry, timestamp)
  File "/Users/austinov/wavefronthq/wavefront-pyformance/wavefront_pyformance/wavefront_reporter.py", line 53, in report_now
    col.collect()
  File "/Users/austinov/wavefronthq/wavefront-pyformance/wavefront_pyformance/runtime_metrics.py", line 180, in collect
    self.collect_io()
  File "/Users/austinov/wavefronthq/wavefront-pyformance/wavefront_pyformance/runtime_metrics.py", line 44, in collect_io
    io_counters = self.process.io_counters()
AttributeError: 'Process' object has no attribute 'io_counters'
```

File: wavefront_pyformance/wavefront_reporter.py (hunk 2 above)
Comment (sjaiswalvmware) [reply wavefront_pyformance/wavefront_reporter.py:27]: @sushantdewan123 Did not touch the interval here. It was as it is.

File: example_runtime_metrics.py (hunk 3 above)
Comment (sjaiswalvmware) [reply example_runtime_metrics.py]: @ustinov It was commented as either Proxy or Direct reporter would be used in real scenario. I have uncommented it to keep the way it is in example.py.

File: wavefront_pyformance/runtime_metrics.py (hunk 4 above)
Comment (sjaiswalvmware) [reply wavefront_pyformance/runtime_metrics.py]: @ustinov Done.

File: README.md (hunk 5 above)
Comment (sjaiswalvmware) [reply README.md]: @ustinov The flag name is changed to enable_runtime_metrics for better understanding.

File: .travis.yml (hunk 6 above)
Comment (sjaiswalvmware) [reply .travis.yml]: @ustinov Done.

File: wavefront_pyformance/runtime_metrics.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                            "process_name": self.pname,
+                            "process_status": self.status}
+
+    def collect_cputimes(self):
+        """Collect CPU Times."""
+        cputimes = self.process.cpu_times()
+        system_cpu = cputimes[1]  # system cpu
+        metric1 = self.registry.gauge("cpu.times",
+                                      tags=self.custom_tags)
+        metric1.set_value(system_cpu)
```
Comment (ustinov) [inline wavefront_pyformance/runtime_metrics.py]: We could set the value right away after retrieving the gauge, e.g.
```
        self.registry.gauge("cpu.times",
                            tags=self.custom_tags).set_value(system_cpu)
```
here and further down

Comment (ustinov) [reply wavefront_pyformance/runtime_metrics.py]: WDYT?

File: wavefront_pyformance/runtime_metrics.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    def collect_memorypercent(self):
+        """Collect Memory in Percentage."""
+        memorypercent = self.process.memory_percent(memtype="rss")
+        metric1 = self.registry.gauge("memory.rss.percent",
+                                      tags=self.custom_tags)
+        metric1.set_value(memorypercent)
+
+    def collect_threads(self):
+        """Collect Threading Metrics."""
+        counter = 0
```
Comment (ustinov) [inline wavefront_pyformance/runtime_metrics.py]: You may use tuple assignment, as you use further in your code
```
counter, alive, daemon = 0, 0, 0
```

File: wavefront_pyformance/runtime_metrics.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        metric2 = self.registry.gauge("thread.daemon",
+                                      tags=self.custom_tags)
+        metric2.set_value(daemon)
+        metric3 = self.registry.gauge("thread.alive",
+                                      tags=self.custom_tags)
+        metric3.set_value(alive)
+
+    def collect_garbage(self):
+        """Collect Garbage Collection Metrics."""
+        (count0, count1, count2) = gc.get_count()
```
Comment (ustinov) [inline wavefront_pyformance/runtime_metrics.py]: Parens are optional

**Comment by ustinov** [comment]
Added @akodali18 as we need this tested in AWS lambda environment before merging

File: wavefront_pyformance/runtime_metrics.py (hunk 7 above)
Comment (sjaiswalvmware) [reply wavefront_pyformance/runtime_metrics.py]: @ustinov Yes, that's better. Made the change.

File: wavefront_pyformance/runtime_metrics.py (hunk 8 above)
Comment (sjaiswalvmware) [reply wavefront_pyformance/runtime_metrics.py]: Done

File: wavefront_pyformance/runtime_metrics.py (hunk 9 above)
Comment (sjaiswalvmware) [reply wavefront_pyformance/runtime_metrics.py]: Done.

**Review by ustinov — APPROVED** [review]
LGTM. Thanks @sjaiswalvmware for addressing all comments

**Review by haosong — APPROVED** [review]
lgtm

**Review by sushantdewan123 — APPROVED** [review]

**Pull request merged**